### PR TITLE
Fix Grover edge cases from Copilot suggestions

### LINKED ISSUE #347: Check and implement suggestions suggested by Copilot in Grover search
- Copilot suggested a couple of suggestion in #346 related to our Grover search implementation. Those were unrelated to that PR.
- Check and implement the valid suggestions by Copilot

### PR DESCRIPTION
**Pull request by DeconBear**
Closes #347
Supersedes #350

  ## Summary
  This PR implements valid Copilot follow-up suggestions for Grover search:

  - Normalize `search_qubits` input in both `grover_oracle` and `grover` (convert iterable inputs to list).
  - Raise `ValueError` when `search_qubits` is empty.
  - Fix the 1-qubit oracle edge case by correctly uncomputing flipped zero bits for all qubit counts.
  - Add/keep regression tests for:
    - empty `search_qubits` in `grover_oracle`
    - empty `search_qubits` in `grover`
    - 1-qubit oracle marking `|0>`
  - Keep the test dims handling fix for overlap checks in multi-solution Grover tests.

  ## Files changed
  - `src/qutip_qip/algorithms/grover.py`
  - `tests/test_grover.py`

  ## Testing
  - Local environment is Python 3.9.7, while this branch uses `|` type annotations that require newer Python.
  - Please rely on CI for full validation on supported Python versions.

### CODE REVIEW

File: src/qutip_qip/algorithms/grover.py [hunk 1]
```diff
@@ -136,6 +141,11 @@ def grover(
     """
     if isinstance(search_qubits, int):
         search_qubits = list(range(search_qubits))
+    else:
+        search_qubits = list(search_qubits)
```
Comment (Mayank447) [inline src/qutip_qip/algorithms/grover.py]:
```suggestion
    elif not isinstance(search_qubits, SequenceLike):
        raise TypeError("search_qubits must be an int or a Sequence of int")
```

You will need do the import `from qutip_qip.typing import SequenceLike`

File: src/qutip_qip/algorithms/grover.py [hunk 2]
```diff
@@ -27,6 +27,11 @@ def grover_oracle(
...
```
Comment (Mayank447) [inline src/qutip_qip/algorithms/grover.py]: Same change as previous.

File: tests/test_grover.py [hunk 3]
```diff
@@ -109,8 +137,13 @@ def test_grover_3_qubit_multiple_targets(self):
         psi_final = U_grover * psi0
 
         # Check probability of measuring EITHER 3 or 5
-        prob_3 = abs(psi_final.overlap(basis(1 << n_qubits, 3))) ** 2  # 1 << N = 2**N
-        prob_5 = abs(psi_final.overlap(basis(1 << n_qubits, 5))) ** 2
+        dims = [[2] * n_qubits, [1] * n_qubits]
+        state_3 = basis(2**n_qubits, 3)
+        state_3.dims = dims
```
Comment (Mayank447) [inline tests/test_grover.py]:
```suggestion
        state_3 = basis(2**n_qubits, 3)
```

File: tests/test_grover.py [hunk 4]
```diff
@@ -109,8 +137,13 @@ def test_grover_3_qubit_multiple_targets(self):
...
+        state_5 = basis(2**n_qubits, 5)
+        state_5.dims = dims
```
Comment (Mayank447) [inline tests/test_grover.py]:
```suggestion
        state_5 = basis(2**n_qubits, 5)
```

**Review by Mayank447 — COMMENTED** [review]
Looks good to me, just minor changes.

**Comment by DeconBear** [comment]
Addressed in f6c816d9

> Looks good to me, just minor changes.

File: src/qutip_qip/algorithms/grover.py [hunk 5]
```diff
@@ -27,6 +27,13 @@ def grover_oracle(
     """
     if isinstance(search_qubits, int):
         search_qubits = list(range(search_qubits))
+    elif not isinstance(search_qubits, SequenceLike):
+        raise TypeError("search_qubits must be an int or a Sequence of int")
+    else:
```
Comment (Mayank447) [inline src/qutip_qip/algorithms/grover.py]: This else statement is no longer required anymore

File: src/qutip_qip/algorithms/grover.py [hunk 6]
```diff
@@ -136,6 +143,13 @@ def grover(
...
```
Comment (Mayank447) [inline src/qutip_qip/algorithms/grover.py]: Same here

File: src/qutip_qip/algorithms/grover.py [hunk 7]
```diff
@@ -29,8 +29,7 @@ def grover_oracle(
         search_qubits = list(range(search_qubits))
     elif not isinstance(search_qubits, SequenceLike):
         raise TypeError("search_qubits must be an int or a Sequence of int")
-    else:
-        search_qubits = list(search_qubits)
+    search_qubits = list(search_qubits)
```
Comment (Mayank447) [inline src/qutip_qip/algorithms/grover.py]: Remove this line completely.

File: src/qutip_qip/algorithms/grover.py [hunk 8]
```diff
@@ -145,8 +144,7 @@ def grover(
...
```
Comment (Mayank447) [inline src/qutip_qip/algorithms/grover.py]: Same as above

**Review by Mayank447 — APPROVED** [review]

File: src/qutip_qip/algorithms/grover.py (hunk 8 above)
Comment (DeconBear) [reply src/qutip_qip/algorithms/grover.py]: Addressed in 37e9f378. Removed the two redundant lines as suggested.

**Comment by Mayank447** [comment]
@BoxiLi you can go ahead with the merge. Btw.. do a squash merge
